labs/tsc.bzl

```diff
@@ -0,0 +1,133 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A docstring"""
+
+load("@build_bazel_rules_nodejs//:providers.bzl", "DeclarationInfo", "NpmPackageInfo", "node_modules_aspect", "run_node")
+load("//internal/common:expand_variables.bzl", "expand_variables")
+load("//internal/linker:link_node_modules.bzl", "module_mappings_aspect")
+
+_ATTRS = {
+    "srcs": attr.label_list(
+        allow_files = [".ts", ".tsx"],
+        mandatory = True,
+    ),
+    "outs": attr.output_list(),
+    "args": attr.string_list(),
+    "declarations": attr.bool(default = False),
+    "sourcemaps": attr.bool(default = False),
+    "tsc": attr.label(
+        default = "@npm//typescript/bin:tsc",
+        executable = True,
+        cfg = "host",
+    ),
+    "tsconfig": attr.label(
+        default = "//:tsconfig.json",
+        allow_single_file = True,
+    ),
+    "deps": attr.label_list(
+        allow_files = True,
+        aspects = [module_mappings_aspect, node_modules_aspect],
+    ),
+}
+
+def _basename_no_ext(f):
+    return f.basename[:-len(f.extension) - 1]
+
+def _expand_locations(ctx, s):
+    # `.split(" ")` is a work-around https://github.com/bazelbuild/bazel/issues/10309
+    # _expand_locations returns an array of args to support $(execpaths) expansions.
+    # TODO: If the string has intentional spaces or if one or more of the expanded file
+    # locations has a space in the name, we will incorrectly split it into multiple arguments
+    targets = ctx.attr.deps + [ctx.attr.tsconfig]
+    return ctx.expand_location(s, targets = targets).split(" ")
+
+def _deps_inputs(ctx):
+    depsets = []
+    for d in ctx.attr.deps:
+        # Only include DefaultInfo files from a dep if there is no DeclarationInfo. If there is a
+        # DeclarationInfo then tsc will rely on the declaration files from that target.
+        if not DeclarationInfo in d:
+            depsets.append(d[DefaultInfo].files)
+    return depset(transitive = depsets)
+
+def _declaration_inputs(ctx):
+    depsets = []
+    for d in ctx.attr.deps:
+        if DeclarationInfo in d:
+            depsets.append(d[DeclarationInfo].transitive_declarations)
+    return depset(transitive = depsets)
+
+def _npm_inputs(ctx):
+    depsets = []
+    for d in ctx.attr.deps:
+        if NpmPackageInfo in d:
+            depsets.append(d[NpmPackageInfo].sources)
+    return depset(transitive = depsets)
+
+def _impl(ctx):
```

Comment:
this looks awesome!

Reply:
I also see this as a real nice example for a rule which not only consumes but produces providers

Reply:
Thanks! I was aiming for a nice clean implementation that was easy to follow. Not done yet as I'd like to pull in the JSModuleInfo provider as well.

Reply:
I also setup snapshot builds for rules_nodejs. https://github.com/gregmagolan/rules_nodejs-builds/tree/labs

Will upstream in the near future and have the snapshots published to https://github.com/bazelbuild/rules_nodejs-builds

Reply:
Cool. I'm pretty happy with that now except for all of the missing docstrings. I'll do something similar for jest_test.

I'm hoping the ngc rule will be as simple as:

```
load(":tsc.bzl", "tsc")

def ngc(**kwargs, ngc = "@npm//@angular/cli/bin:ngc"):
    """A docstring"""
    tsc(
        ngc = ngc,
        **kwargs
    )
```

Reply:
We might be able to get around the VE use case with Ivy on by default on v9.
But being able to swap out compilers like that is awesome!
It also means people would be able to swap out for custom ts transform pipelines as well